jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/Callback.java

```diff
@@ -592,4 +641,25 @@ public Completable compose(Consumer<Completable> consumer)
             return completable;
         }
     }
+
+    /**
+     * Invoke a callback failure, handling any {@link Throwable} thrown
+     * by adding the passed {@code failure} as a suppressed with
+     * {@link ExceptionUtil#addSuppressedIfNotAssociated(Throwable, Throwable)}.
+     * @param callback The callback to fail
+     * @param failure The failure
+     * @throws RuntimeException If thrown, will have the {@code failure} added as a suppressed.
+     */
+    static void failed(Callback callback, Throwable failure)
+    {
+        try
+        {
+            callback.failed(failure);
+        }
+        catch (Throwable t)
+        {
+            ExceptionUtil.addSuppressedIfNotAssociated(t, failure);
+            throw t;
```

Comment:
`Callback.failed()` and `Callback.succeeded()` are supposed to never throw, aren't they?

I like this new static helper, which makes it easier to ensure that (maybe there should be the equivalent for succeed() and we should always use them?) but I'm not sure rethrowing the exception is the right thing to do here.

Reply:
@lorban I'm not sure if failed should or should not throw, it is not clear in the contract.    What is clear is that sometimes they do through and we mostly try to handle that.
However, whilst handling these throws, we lose the information of what the original exception that was passed into `failed(Throwable)` was.

So the only intention of this change is to record the original failure as a a suppressed exception on the newly thrown exception.  Hence we still throw the exception as we have done before.

There is no equivalent for `succeeded` as there is no original exception.

Reply:
Yeah... I've had several goes at that.... it pretty quickly devolves into reinventing `CompletableFuture`... but perhaps there is a good middle ground.